osp/core/ontology/parser/yml/parser.py

```diff
@@ -405,8 +405,12 @@ def _add_attribute(self, class_iri, attribute_iri, default):
             self._graph.add((class_iri, rdflib_cuba._default, bnode))
             self._graph.add(
                 (bnode, rdflib_cuba._default_attribute, attribute_iri))
+            # Find datatype of attribute.
+            data_type = self._graph.value(attribute_iri, RDFS.range)
+            lexicalized_default = str(Literal(default, datatype=data_type))
             self._graph.add(
-                (bnode, rdflib_cuba._default_value, Literal(default)))
+                (bnode, rdflib_cuba._default_value,
+                 Literal(lexicalized_default, datatype=data_type)))
```

Comment:
In the previous one I would have to check whether it makes sense, but here it is the case.

I know it is very weird and I thought the same when I learned I had to do this, and I agree that it is ugly.

It boils down to the fact that running

```
from rdflib import Literal, XSD
x = Literal(8, datatype=XSD.string)
print(type(x.toPython()) == int)
```

prints `True`.

Have a look at the lexicalizer stuff on `datatypes.py`. Basically in RDFLib you can register functions to convert a python object to a string (lexicalizer) and to convert a string to a literal (constructor I think was the name). This is necessary because in RDF everything is serialized as a string.

When a literal is created, it saves the original Python object, so that when you call `.toPython()` it returns such object (that's why the code above prints `True`). However, if the literal came from an imported RDF file or from a string, then calling `.toPython()` actually calls the constructor.  I think this behaviour makes no sense, and `.toPython()` should always call the constructor. If you have the time and are sure that you really understand the issue you can report this to RDFLib's developers.

So the point of calling it twice is:
- `lexicalized_default = str(Literal(default, datatype=data_type))` -> Serializes the Python object as a string compatible with RDF.
- `Literal(lexicalized_default, datatype=data_type)` -> Creates a new literal. Since this literal came from a string, calling `toPython()` will call the constructor and give the right type of python object.

If I do not do this, then when the Literal is fetched from the graph from other part of the code and `toPython()` is called, I will get the original object. This breaks other stuff like the sqlite wrapper, plus gives unexpected results. If you're interested in this stuff and this explanation is not enough we can also have a meeting.

Reply:
Maybe we can discuss this briefly in the next meeting. Not sure I understand the problem.
An example of an unexpected behaviour might help.